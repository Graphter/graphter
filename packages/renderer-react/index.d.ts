import ListRenderer from './list-renderer';
export { ListRenderer };
