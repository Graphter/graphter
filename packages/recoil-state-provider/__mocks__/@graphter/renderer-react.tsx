import { NodeRendererRegistration } from "@graphter/core";
import React from "react";

export const nodeRendererStore = {
  get: jest.fn()
}

export const pathConfigStore = {
  get: jest.fn(),
  set: jest.fn()
}

export const NodeValidationProvider = jest.fn(({ children }: any) => <div>{children}</div>)
export const NodeDataProvider = jest.fn(({ children }: any) => <div>{children}</div>)