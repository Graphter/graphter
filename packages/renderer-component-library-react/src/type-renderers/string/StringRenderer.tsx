import React, { useMemo } from "react";
import { PropertyTypeRendererProps } from "malle-renderer-react";
import s from './StringRenderer.pcss'

export default function StringRenderer(
  {
    propertyConfig,
    modelConfig,
    propData,
    modelData,
    setPropDataValue,
    setModelDataValue,
    validationResults
  }: PropertyTypeRendererProps
){
  return (
    <div>
      <label htmlFor={propertyConfig.id}>{propertyConfig.name}</label>
      {propertyConfig.description && <p>{propertyConfig.description}</p>}
      <input
        type='text'
        id={propertyConfig.id}
        defaultValue={propData}
        data-testid='string-input'
        onChange={(e) => {
          setPropDataValue && setPropDataValue(e.currentTarget.value);
        }} />
      {validationResults && validationResults.map((result, i) => (
        <div className={s.error} key={i}>{result.errorMessage}</div>
      ))}
    </div>
  );
}