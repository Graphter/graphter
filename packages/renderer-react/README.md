# Graphter React renderer

A React framework that, when combined with node renderers such as those provided in the [Component Lib](../renderer-component-library-react/README.md) package, allows humans to manage arbitrary data structures.