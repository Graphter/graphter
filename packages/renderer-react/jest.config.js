module.exports = {
  preset: 'ts-jest',
  testEnvironment: 'jsdom',
  moduleNameMapper: {
    "\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "<rootDir>/__mocks__/fileMock.ts",
    "\\.(pcss|css|less)$": "identity-obj-proxy"
  },
  setupFilesAfterEnv: [ './setupTests.ts' ],
  automock: true,
  unmockedModulePathPatterns: [],
  snapshotSerializers: [
    'enzyme-to-json'
  ],
  testMatch: ['<rootDir>/**/*.test.ts*'],
  testPathIgnorePatterns: ['node_modules'],
};