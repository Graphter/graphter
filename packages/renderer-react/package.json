{
  "name": "malle-renderer-react",
  "version": "1.0.0",
  "description": "",
  "main": "dist/index.js",
  "module": "dist/index.es.js",
  "files": [
    "dist"
  ],
  "types": "dist/index.d.ts",
  "scripts": {
    "test": "cross-env CI=1 react-scripts test --env=jsdom",
    "test:watch": "react-scripts test --env=jsdom",
    "build": "rollup -c",
    "start": "rollup -c -w",
    "prepare": "npm run build"
  },
  "author": "",
  "license": "ISC",
  "peerDependencies": {
    "react": "^16.13.1",
    "react-dom": "^16.13.1"
  },
  "dependencies": {},
  "devDependencies": {
    "react": "^16.13.1",
    "react-dom": "^16.13.1",
    "@babel/core": "^7.9.0",
    "@babel/plugin-proposal-export-default-from": "^7.8.3",
    "@babel/preset-env": "^7.9.0",
    "@babel/preset-react": "^7.9.1",
    "@babel/preset-stage-0": "^7.8.3",
    "@svgr/rollup": "^5.3.0",
    "@types/react": "^16.9.25",
    "eslint": "^6.8.0",
    "react-scripts": "^3.4.1",
    "rollup": "^2.2.0",
    "rollup-plugin-babel": "^4.4.0",
    "rollup-plugin-commonjs": "^10.1.0",
    "rollup-plugin-node-resolve": "^5.2.0",
    "rollup-plugin-peer-deps-external": "^2.2.2",
    "rollup-plugin-postcss": "^2.5.0",
    "rollup-plugin-typescript2": "^0.26.0",
    "rollup-plugin-url": "^3.0.1",
    "typescript": "^3.8.3"
  }
}
