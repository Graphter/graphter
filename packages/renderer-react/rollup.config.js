import babel from 'rollup-plugin-babel'
import commonjs from 'rollup-plugin-commonjs'
import external from 'rollup-plugin-peer-deps-external'
import postcss from 'rollup-plugin-postcss'
import resolve from 'rollup-plugin-node-resolve'
import url from 'rollup-plugin-url'
import typescriptPlugin from 'rollup-plugin-typescript2'
import typescript from 'typescript';
import svgr from '@svgr/rollup'

import pkg from './package.json'

const externals = [
  ...Object.keys(pkg.peerDependencies),
  ...Object.keys(pkg.dependencies)
];
console.log('externals', externals);

export default {
  input: 'src/index.ts',
  output: [
    {
      file: pkg.main,
      format: 'cjs',
      sourcemap: true
    },
    {
      file: pkg.module,
      format: 'es',
      sourcemap: true
    }
  ],
  plugins: [
    // For debugging: { load(id) {console.log(id)} },
    external(externals),
    typescriptPlugin({
      typescript
    }),
    postcss({
      modules: true
    }),
    url(),
    svgr(),
    babel({
      exclude: 'node_modules/**',
      plugins: [ '@babel/plugin-proposal-export-default-from' ]
    }),
    resolve(),
    commonjs({
      namedExports: {
        'recoil': [ 'RecoilRoot', 'useRecoilState', 'useRecoilValue', 'useRecoilValueLoadable', 'useRecoilCallback', 'atom', 'selector' ]
      }
    })
  ]
}
