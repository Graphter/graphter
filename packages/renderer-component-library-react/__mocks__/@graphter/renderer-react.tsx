import React, { useState } from "react";
import { PathSegment } from "@graphter/core";

export const useArrayNodeData = jest.fn(() => {
  return {
    childIds: [],
    removeItem: jest.fn(),
    commitItem: jest.fn()
  }
})

export const useNodeData = jest.fn(() => {
  return [
    null,
    () => {}
  ]
})

export const useNodeValidation = jest.fn().mockResolvedValue([])

export const nodeRendererStore = {
  get: jest.fn((nodeType: string) => ({
    type: 'string',
    Renderer: ({originalNodeData}: any) => {
      const [ value, setValue ] = useState(originalNodeData || '')
      return (
        <input value={value} onChange={e => setValue(e.target.value)} data-testid='string-renderer' />
    )
    }
  }))
}
export const createDefault = jest.fn()

export const pathConfigStore = {
  get: jest.fn(),
  set: jest.fn()
}

export const setupNodeRenderer = jest.fn((renderer: any) => renderer)

export {}