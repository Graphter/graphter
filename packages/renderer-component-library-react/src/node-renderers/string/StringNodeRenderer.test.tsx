import React from "react";
import { act, render, fireEvent } from '@testing-library/react';
import * as graphterRenderer from "@graphter/renderer-react";
import StringRenderer from "./StringNodeRenderer";

jest.spyOn(graphterRenderer, 'useNodeData').mockImplementation(() => [ null, () => {} ])
jest.spyOn(graphterRenderer, 'useNodeValidation').mockImplementation(() => [] )

const useNodeDataMock = graphterRenderer.useNodeData as  jest.Mock<any>
const useNodeValidationMock = graphterRenderer.useNodeValidation as  jest.Mock<any>

describe(`<StringRenderer />`, () => {
  afterEach(() => {
    useNodeDataMock.mockClear()
  })
  it(`should render correctly`, () => {
    useNodeDataMock.mockReturnValue([ 'some-data', () => {} ])
    const { container } = render(<StringRenderer
      config={{
        id: 'name',
        name: 'Name',
        description: 'The name',
        type: 'string',
        default: 'Unnamed'
      }}
      originalNodeData={'Foo Bar'}
      committed={true}
      path={['/']}
    />);
    expect(container).toMatchSnapshot();
  });
  it('should render default data when new', () => {
    useNodeDataMock.mockReturnValue([ 'The default value', () => {} ])
    const { getByDisplayValue } = render(<StringRenderer
      config={{
        id: 'name',
        name: 'Name',
        description: 'The name',
        type: 'string',
        default: 'The default value'
      }}
      originalNodeData={undefined}
      committed={true}
      path={['/']}
    />);
    expect(useNodeDataMock).toHaveBeenCalled()
    expect(useNodeDataMock.mock.calls[0][2]).toBe('The default value')
  })
  it('should use the data provider for data', () => {
    useNodeDataMock.mockReturnValue([ 'The data provider value', () => {} ])
    const { queryByDisplayValue } = render(<StringRenderer
      config={{
        id: 'name',
        name: 'Name',
        description: 'The name',
        type: 'string',
        default: 'The default value'
      }}
      originalNodeData={'The original value'}
      committed={true}
      path={['/']}
    />);
    expect(queryByDisplayValue('The data provider value')).toBeInTheDocument()
  })
  it('should show validation results when touched', () => {
    useNodeDataMock.mockReturnValue([ 'The data provider value', () => {} ])
    useNodeValidationMock.mockReturnValue([
      {
        errorMessage: 'Some validation error message'
      }
    ])
    const { getByDisplayValue, queryByText } = render(<StringRenderer
      config={{
        id: 'name',
        name: 'Name',
        description: 'The name',
        type: 'string',
        default: 'The default value'
      }}
      originalNodeData={'The original value'}
      committed={true}
      path={['/']}
    />);
    act(() => {
      fireEvent.change(getByDisplayValue('The data provider value'), { target: { value: '' }})
    })
    expect(queryByText('Some validation error message')).toBeInTheDocument()
  })
  it('should not show validation results when not touched', () => {
    useNodeDataMock.mockReturnValue([ 'The data provider value', () => {} ])
    useNodeValidationMock.mockReturnValue([
      {
        errorMessage: 'Some validation error message'
      }
    ])
    const { queryByText } = render(<StringRenderer
      config={{
        id: 'name',
        name: 'Name',
        description: 'The name',
        type: 'string',
        default: 'The default value'
      }}
      originalNodeData={'The original value'}
      committed={true}
      path={['/']}
    />);
    expect(queryByText('Some validation error message')).not.toBeInTheDocument()
  })
})

export {};