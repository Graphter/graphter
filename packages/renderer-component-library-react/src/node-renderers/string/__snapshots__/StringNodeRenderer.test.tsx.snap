// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StringNodeRenderer /> should render correctly 1`] = `
<div>
  <input
    class="input"
    data-nodepath="/"
    data-nodetype="string"
    type="text"
    value="some-data"
  />
</div>
`;

exports[`<StringRenderer /> should render correctly 1`] = `
<div>
  <input
    class="input"
    data-nodepath="/"
    data-nodetype="string"
    type="text"
    value="some-data"
  />
</div>
`;
